editor/tree_search.h

```diff
@@ -0,0 +1,121 @@
+ */
+
+#ifdef TOOLS_ENABLED
+
+#ifndef TREE_SEARCH_H
+#define TREE_SEARCH_H
+
+#include "../bt/tasks/bt_task.h" // for tree item parsing
+
+#ifdef LIMBOAI_GDEXTENSION
+#include <godot_cpp/classes/check_box.hpp>
+#include <godot_cpp/classes/h_flow_container.hpp>
+#include <godot_cpp/classes/label.hpp>
+#include <godot_cpp/classes/line_edit.hpp>
+#include <godot_cpp/classes/tree.hpp>
+
+#endif // LIMBOAI_GDEXTENSION
+
+#ifdef LIMBOAI_MODULE
+#include "scene/gui/check_box.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/label.h"
+#include "scene/gui/line_edit.h"
+#include "scene/gui/tree.h"
+#endif // LIMBOAI_MODULE
+
+using namespace godot;
+
+enum TreeSearchMode {
+	HIGHLIGHT = 0,
+	FILTER = 1
+};
+
+class TreeSearchPanel : public HFlowContainer {
+	GDCLASS(TreeSearchPanel, HFlowContainer)
+private:
+	Button *toggle_button_filter_highlight;
+	Button *close_button;
+	Label *label_filter;
+	LineEdit *line_edit_search;
+	CheckBox *check_button_filter_highlight;
+	void _initialize_controls();
+	void _initialize_close_callbacks();
+	void _add_spacer(float width_multiplier = 1.f);
+
+	void _on_draw_highlight(TreeItem *p_item, Rect2 p_rect);
+	void _emit_text_submitted(const String &p_text);
+	void _emit_update_requested();
+	void _notification(int p_what);
+
+protected:
+	static void _bind_methods();
+
+public:
+	TreeSearchMode get_search_mode();
+	String get_text();
+	void show_and_focus();
+	TreeSearchPanel();
+	bool has_focus();
+};
+
+class TreeSearch : public RefCounted {
+	GDCLASS(TreeSearch, RefCounted)
+private:
+	struct StringSearchIndices {
+		// initialize to opposite bounds.
+		int lower = -1;
+		int upper = -1;
+
+		bool hit() {
+			return 0 <= lower && lower < upper;
+		}
+	};
+
+	Tree *tree_reference;
+
+	Vector<TreeItem *> ordered_tree_items;
+	Vector<TreeItem *> matching_entries;
+	HashMap<TreeItem *, int> number_matches;
+	HashMap<TreeItem *, Callable> callable_cache;
+
+	void _filter_tree(const String &p_search_mask);
+	void _highlight_tree(const String &p_search_mask);
+	void _draw_highlight_item(TreeItem *p_tree_item, Rect2 p_rect, String p_search_mask, Callable p_parent_draw_method);
+	void _update_matching_entries(const String &p_search_mask);
+	void _update_ordered_tree_items(TreeItem *p_tree_item);
+	void _update_number_matches();
+
+	Vector<TreeItem *> _find_matching_entries(TreeItem *p_tree_item, const String &p_search_mask, Vector<TreeItem *> &p_buffer);
+	StringSearchIndices _substring_bounds(const String &p_searchable, const String &p_search_mask) const;
+
+	void _select_item(TreeItem *p_item);
+	void _select_first_match();
+	void _select_next_match();
+
+	template <typename T>
+	bool _vector_has_bsearch(Vector<T *> p_vec, T *element);
+
+protected:
+	static void _bind_methods() {}
+
+public:
+	// we will add everything from TaskTree.h
+	void update_search(Tree *p_tree);
+	void on_item_edited(TreeItem *p_item);
+	TreeSearchPanel *search_panel;
```

Comment:
So you are proposing to construct the `TreeSearchPanel` in `TaskTree` and calling `TreeSearch(tree_search_panel)`?

I wouldn't even though of that. Connecting the `update_requested_signal` causes no issue then, because we have access.

The TaskTree would then maintain privately maintain TreeSearchPanel * tree_search_panel for 
```
void TaskTree::tree_search_show_and_focus() {
	tree_search->search_panel->show_and_focus();
}
```
(which is called by the misc entry).
Correct?

Reply:
It could be owned either by the TaskTree or by the Editor, whatever makes more sense. Haven't read the code in fullest yet.

Reply:
[47706e9](https://github.com/limbonaut/limboai/pull/229/commits/47706e9480b6fc831e171f9a6c39ca0d7b6d988c) Should address this.